Fix: crash due to incomplete deity/company mode precondition checks

rubidium42 opened the pull request:
## Motivation / Problem

Inconsistent/missing checks for being in deity/company mode. For example, for setting the industry text there is no explicit check for it being called from deity mode. If called from company mode the call will eventually fail in the command processing, instead of earlier.

Furthermore there is a potential crash to desktop when the company mode becomes invalid.


## Description

The crashes to desktop are due to the company of the CompanyMode becoming invalid after a merge or bankruptcy, and then the removed company is still being used in getters and commands. This causes dereferences of nullptr.
This is solved by checking in `ScriptCompanyMode::IsValid` for the company of the company mode still being a valid company.

All cases where `ScriptObject::GetCompany` is used, including all commands, are suspect for this potential issue of using an invalid company. So, add a precondition check for either the right company mode or deity mode being active to all those functions.

Furthermore a script has been made that finds such cases, and this is integrated in a workflow that checks for this during builds on github.

Finally the last error for all these precondition checks is changed to `ERR_PRECONDITION_INVALID_COMPANY` instead of the less specific `ERR_PRECONDITION_INVALID`.


## Limitations

None as far as I know.


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

rubidium42 committed:
Change: upon an invalid company mode, set ERR_PRECONDITION_INVALID_COMPANY instead of ERR_PRECONDITION_INVALID

glx22 commented on src/script/api/script_depotlist.cpp:
```diff
@@ -16,6 +16,7 @@
 
 ScriptDepotList::ScriptDepotList(ScriptTile::TransportType transport_type)
 {
+	EnforceDeityOrCompanyModeValid();
```
This is wrong for every builds, but only MSVC warns.

rubidium42 replied:
I know, I've fixed it locally.

I'm working on splitting this PR in smaller, hopefully easier to review PRs as it has become a bit of a mess in this branch. Like first a PR for the trivial refactor, and then update this PR to add all the missing ones with logic to prevent them going missing in the future.

rubidium42 committed:
Add: company mode enforcement checks to functions with command or company access

Command functions are those that call ScriptObject::Command, and functions
with company access are any that call ScriptObject::GetCompany. This is a bit
over-protective, but having the check everywhere makes it easier to validate
that no check is missing automatically instead of by review.

rubidium42 committed:
Add: deity mode enforcement checks to functions with command or company access

Command functions are those that call ScriptObject::Command, and functions
with company access are any that call ScriptObject::GetCompany. This is a bit
over-protective, but having the check everywhere makes it easier to validate
that no check is missing automatically instead of by review.

rubidium42 committed:
Add: precondition checks to functions that work with both valid company and deity

These are functions that either use ScriptObject::Command or ScriptObject::GetCompany.
This is a bit over-protective, but having the check everywhere makes it easier to
validate that no check is missing automatically instead of by review.

At this moment these checks will not do anything useful, as either IsValid or
IsDeity from ScriptCompanyMode returns true, but that will change later.

rubidium42 committed:
Fix: crashes to desktop from game scripts when companies disappear

When a game script is in company mode, it pretends to be another company. When
that company disappear (bankruptcy/merger), the game script still uses that
company and it keeps calling functions as if it is that company.

For example, ScriptEngine::IsBuildable internally dereferences Company without
checks, causing a null dereference for any ScriptEngine function when called
from a company scope of a company that has disappeared.

Guard against this by extending the ScriptCompanyScope::IsValid check to also
check for the company still being active.

rubidium42 committed:
Add: workflow and script for checking missing mode enforcements

SamuXarick commented on src/script/api/script_airport.cpp:
```diff
@@ -91,24 +91,26 @@
 
 /* static */ int32 ScriptAirport::GetNumHangars(TileIndex tile)
 {
+	EnforceDeityOrCompanyModeValid(false);
```
```suggestion
	EnforceDeityOrCompanyModeValid(-1);
```

SamuXarick commented on src/script/api/script_airport.cpp:
```diff
@@ -91,24 +91,26 @@
 
 /* static */ TileIndex ScriptAirport::GetHangarOfAirport(TileIndex tile)
 {
+	EnforceDeityOrCompanyModeValid(false);
```
```suggestion
	EnforceDeityOrCompanyModeValid(INVALID_TILE);
```

SamuXarick commented on src/script/api/script_sign.cpp:
```diff
@@ -71,6 +74,7 @@
 {
 	CCountedPtr<Text> counter(name);
 
+	EnforceDeityOrCompanyModeValid(false);
```
```suggestion
	EnforceDeityOrCompanyModeValid(INVALID_SIGN);
```

SamuXarick commented on src/script/api/script_town.cpp:
```diff
@@ -277,11 +281,12 @@
-	EnforcePrecondition(false, size != TOWN_SIZE_LARGE || ScriptObject::GetCompany() == OWNER_DEITY);
-	if (ScriptObject::GetCompany() == OWNER_DEITY || _settings_game.economy.found_town == TF_CUSTOM_LAYOUT) {
+	EnforcePrecondition(false, ScriptCompanyMode::IsDeity() || size != TOWN_SIZE_LARGE);
+	if (ScriptCompanyMode::IsDeity() || _settings_game.economy.found_town == TF_CUSTOM_LAYOUT) {
```
You switched the evaluation order. Does it matter?

rubidium42 replied:
No, it was done for consistency's sake with the check three lines before and to be consistent with the `@pre` documentation.

LordAro commented on .github/script-missing-mode-enforcement.py:
```diff
@@ -0,0 +1,66 @@
+		dangerousFunction = False
+		line = reader.readline()
+
+		while line != '':  # The EOF char is an empty string
```
`for line in reader:`

TrueBrain commented on .github/script-missing-mode-enforcement.py:
```diff
@@ -0,0 +1,62 @@
+import sys
+
+def checkModeEnforcement(path):
+	errors = []
```
It is really uncommon for Python scripts to use tabs; although fully allowed, it is uncommon.

The other scripts for OpenTTD we have been writing with the more default "4 spaces". Maybe good to follow that trend here too?

TrueBrain commented on .github/script-missing-mode-enforcement.py:
```diff
@@ -0,0 +1,62 @@
+			if re.match("^[^\t].*\w::\w", line):
+				modeEnforcementFound = False
+				dangerousFunction = False
+				currentFunction = line
```
For ease of reading, mind doing a `continue` here, and dropping the `else`?

The endless deeper indentation makes for hard-to-read code :)

TrueBrain commented on .github/script-missing-mode-enforcement.py:
```diff
@@ -0,0 +1,62 @@
+
+				if re.match(".*(ScriptObject::Command|ScriptObject::GetCompany).*", line):
+					# Dangerous function found
+					dangerousFunction = True
```
Should the above cases also contain a `continue`? I assume they are not expected to happen on the same line?

TrueBrain commented on .github/script-missing-mode-enforcement.py:
```diff
@@ -0,0 +1,62 @@
+
+				# Line with only a closing bracket. That looks like the end of a function, so check for the dangerous function without mode enforcement
+				if re.match("^}$", line) and dangerousFunction and not modeEnforcementFound:
+					errors.append("%s: %s" % (path, currentFunction.rstrip('\n').replace("/* static */ ", "")))
```
old-style strings are not used that commonly anymore. Suggestion:
```python
function_name = currentFunction.rstrip('\n').replace("/* static */ ", "")
errors.append(f"{path}: {function_name}")
```

TrueBrain commented on .github/script-missing-mode-enforcement.py:
```diff
@@ -0,0 +1,62 @@
+import re
+import sys
+
+def checkModeEnforcement(path):
```
Very nitpicking, feel free to ignore: Python normally uses this_way_of_making_function and similar for this_way_of_variables. Camelcasing really isn't used commonly.

TrueBrain commented on .github/script-missing-mode-enforcement.py:
```diff
@@ -0,0 +1,62 @@
+				dangerousFunction = False
+				currentFunction = line
+			else:
+				if re.match("\t(EnforceDeityMode|EnforceCompanyModeValid|EnforceDeityOrCompanyModeValid|EnforceDeityOrCompanyModeValid_Void)\\(", line):
```
It is wise to use `r""` for regular expressions. This is a "raw" string, and avoids weird things like escaping etc. Not required, just so you are aware that exists.

TrueBrain commented on .github/script-missing-mode-enforcement.py:
```diff
@@ -0,0 +1,62 @@
+	for path in sorted(glob.glob("src/script/api/*.cpp")):
+		# Skip a number of files that yield only false positives
+		if not path.endswith(("script_object.cpp", "script_companymode.cpp", "script_controller.cpp", "script_game.cpp")):
+			errors = errors + checkModeEnforcement(path)
```
The python way is: `errors.extend(checkModeEnforcement(path))` :)

TrueBrain commented on .github/script-missing-mode-enforcement.py:
```diff
@@ -0,0 +1,62 @@
+def main():
+	errors = []
+	for path in sorted(glob.glob("src/script/api/*.cpp")):
+		# Skip a number of files that yield only false positives
```
Comment is a bit deceiving. You don't actually skip ;) What might be easier to read is to do what the comment says:

```python
if path.endswith(("script_object.cpp", "script_companymode.cpp", "script_controller.cpp", "script_game.cpp")):
    continue

errors.extend(checkModeEnforcement(path))
```

100% absolutely nitpicking, so feel free to say: nah, I'm good.

TrueBrain commented on .github/script-missing-mode-enforcement.py:
```diff
@@ -0,0 +1,62 @@
+		for error in errors:
+			print(error)
+		sys.exit(1)
+	else:
```
no need for an `else` after a `sys.exit` :)

TrueBrain commented on .github/script-missing-mode-enforcement.py:
```diff
@@ -0,0 +1,62 @@
+	if errors:
+		print("Mode enforcement was expected in the following files/functions:")
+		for error in errors:
+			print(error)
```
`print("\n".join(errors))`, if you like. Again, matter of opinion, just this join way is more pythonic :D

TrueBrain commented on .github/script-missing-mode-enforcement.py:
```diff
@@ -0,0 +1,62 @@
+
+
+if __name__ == "__main__":
+    main()
```
These days we use a tool called `black` to reformat our Python code. It basically applies the default coding style of Python to a file (https://peps.python.org/pep-0008/ and others). This saves you a lot of time and annoyance.

`pip install black && black -l 120 <filename>` is what we run in all other Python projects we build for OpenTTD. It fixes a few of my comments.
